lib/kb_Metrics/metricsdb_controller.py

```diff
@@ -26,279 +27,190 @@ def _config_str_to_list(self, list_str):
+    def _map_narrative(self, wsid, ws_narrs):
+        """
+        get the narrative name and version
+        """
+        n_name = ''
+        n_obj = '0'
+        for ws in ws_narrs:
+            if str(ws['workspace_id']) == str(wsid):
+                n_name = ws['name']
+                w_meta = ws['meta']
+                for w_m in w_meta:
+                    if w_m['k'] == 'narrative':
+                        n_obj = w_m['v']
+                    elif w_m['k'] == 'narrative_nice_name':
+                        n_name = w_m['v']
+                break
+        return (n_name, n_obj)
 
-    def update_user_info(self, requesting_user, params, token):
+    def _update_user_info(self, requesting_user, params, token):
         """
         update user info
         If match not found, insert that record as new.
         """
-        if not self.is_metrics_admin(requesting_user):
-            raise ValueError('You do not have permission to invoke this action.')
+        if not self._is_metrics_admin(requesting_user):
+            raise ValueError('You do not have permission to '
+                             'invoke this action.')
 
-        params = self.process_parameters(params)
-        # TODO: set the param['minTime'] and param['maxTime'] to a given time window,
-        # such as most current 24 hours instead of 48 hours as for others.
+        params = self._process_parameters(params)
         auth2_ret = self.metrics_dbi.aggr_user_details(
             params['user_ids'], params['minTime'], params['maxTime'])
-        updData = 0
+        upDated = 0
+        upSerted = 0
         if len(auth2_ret) == 0:
             print("No user records returned for update!")
-            return updData
+            return (0, 0)
 
-        print('Retrieved {} user record(s)'.format(len(auth2_ret)))
+        print('Retrieved {} user record(s) for update!'.format(len(auth2_ret)))
         idKeys = ['username', 'email']
         dataKeys = ['full_name', 'signup_at', 'last_signin_at', 'roles']
         for u_data in auth2_ret:
-            filterByKey = lambda keys: {x: u_data[x] for x in keys}
-            idData = filterByKey(idKeys)
-            userData = filterByKey(dataKeys)
+            idData = {x: u_data[x] for x in idKeys}
+            userData = {x: u_data[x] for x in dataKeys}
             isKbstaff = 1 if idData['username'] in self.kbstaffList else 0
             update_ret = self.metrics_dbi.update_user_records(
                 idData, userData, isKbstaff)
-        updData += update_ret.raw_result['nModified']
-
-        return updData
-
-    def update_daily_activities(self, requesting_user, params, token):
+            if update_ret.raw_result['updatedExisting']:
+                upDated += update_ret.raw_result['nModified']
+            elif update_ret.raw_result.get('upserted'):
+                upSerted += 1
+        print('updated {} and upserted {} users.'.format(upDated, upSerted))
+        return (upDated, upSerted)
```

Comment:
I am not sure why you want to return a structure instead of just returning upDated + upSerted.